lib/src/native/realm_core.dart

```diff
@@ -500,52 +502,59 @@ class _RealmCore {
 
     if (syncError is ClientResetError) {
       syncConfig.clientResetHandler.onManualReset?.call(syncError);
-    } else {
-      syncConfig.syncErrorHandler(syncError);
+      return;
     }
+
+    syncConfig.syncErrorHandler(syncError);
   }
 
-  static void _syncBeforeResetCallback(Object userdata, Pointer<shared_realm> realmHandle, Pointer<Void> unlockCallbackFunc) {
-    final syncConfig = userdata as FlexibleSyncConfiguration;
-    final beforeResetCallback = syncConfig.clientResetHandler.onBeforeReset;
-    if (beforeResetCallback == null) {
-      _invokeNativeFunction(unlockCallbackFunc);
-      return;
+  static void _guardSynchCallback(FutureOr<void> Function() callback, Pointer<Void> unlockCallbackFunc) async {
+    bool success = true;
+    try {
+      await callback();
+    } catch (error) {
+      success = false;
+      _realmLib.realm_register_user_code_callback_error(error.toPersistentHandle());
+    } finally {
+      _realmLib.realm_dart_invoke_unlock_callback(success, unlockCallbackFunc);
     }
-    final realm = RealmInternal.getUnowned(syncConfig, RealmHandle._unowned(realmHandle));
-    _continueWhenComplete(() => beforeResetCallback(realm), unlockCallbackFunc);
+  }
+
+  static void _syncBeforeResetCallback(Object userdata, Pointer<shared_realm> realmHandle, Pointer<Void> unlockCallbackFunc) {
+    _guardSynchCallback(() async {
+      final syncConfig = userdata as FlexibleSyncConfiguration;
+      var beforeResetCallback = syncConfig.clientResetHandler.onBeforeReset!;
```

Comment:
I really prefer to have the same check here:
```dart
 if (beforeResetCallback== null) {
        return Future<void>.value();
      }
```
If we throw if for some reason this callback is null then we will always fail the automatics recovery and discard, just because there is no attached callback.

Reply:
this should fail and silence the error. This callback should never be called with onBeforeReset null

Reply:
I don't understand your point. The callback that the users set is clientResetHandler.onBeforeReset but the callback that we pass to core is _syncBeforeResetCallback eventually. I know we have a check while we are creating the config, but should we rely on it? I think that it is better and safety to have the correct check here. If the user hadn't provide a callback we shouldn't break the recover and discard to continue.